File: services/core/repo/mongo/model/timetracking.go
package mongomodel

import (
	"time"

	mongodb "tenkhours/pkg/db/mongo"
	"tenkhours/services/core/entity"

	"github.com/samber/lo"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type TimeTracking struct {
	*mongodb.BaseEntity `                               bson:",inline"`
	CharacterOID        primitive.ObjectID  `json:"characterID,omitempty"   bson:"character_id"`
	CategoryOID         *primitive.ObjectID `json:"categoryID,omitempty"    bson:"category_id"`
	ReferenceOID        *primitive.ObjectID `json:"referenceID,omitempty"   bson:"reference_id"`
	ReferenceType       *entity.EntityType  `json:"referenceType,omitempty" bson:"reference_type"`
	Timestamp           time.Time           `json:"timestamp"               bson:"timestamp"`
	Duration            int                 `json:"duration"                bson:"duration"`
}

func (t *TimeTracking) CharacterID(id string) {
	t.CharacterOID = mongodb.ToObjectID(id)
}

func (t *TimeTracking) CategoryID(id *string) {
	if id != nil {
		t.CategoryOID = lo.ToPtr(mongodb.ToObjectID(*id))
	}
}

func (t *TimeTracking) ReferenceID(id *string) {
	if id != nil {
		t.ReferenceOID = lo.ToPtr(mongodb.ToObjectID(*id))
	}
}

File: services/penk/src/utils/grpc/proto/core/TimeTracking.ts
// Original file: ../../proto/core/timetracking_message.proto

import type {
  EntityType as _core_EntityType,
  EntityType__Output as _core_EntityType__Output,
} from "./EntityType";
import type { Long } from "@grpc/proto-loader";

export interface TimeTracking {
  id?: string;
  characterId?: string;
  categoryId?: string;
  referenceId?: string;
  referenceType?: _core_EntityType;
  timestamp?: number | string | Long;
  duration?: number | string | Long;
  _categoryId?: "categoryId";
  _referenceId?: "referenceId";
  _referenceType?: "referenceType";
}

export interface TimeTracking__Output {
  id: string;
  characterId: string;
  categoryId?: string;
  referenceId?: string;
  referenceType?: _core_EntityType__Output;
  timestamp: string;
  duration: string;
  _categoryId: "categoryId";
  _referenceId: "referenceId";
  _referenceType: "referenceType";
}
